File: apps/backend/src/db/story_store.go
// apps/backend/src/db/story_store.go
package db

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/lucas/oraculo/apps/backend/src/domain"
)

// StoryStore provides CRUD operations for stories backed by SQLite.
// All operations are scoped by epicID.
type StoryStore struct {
	db *DB
}

// NewStoryStore returns a StoryStore that uses the given database connection.
func NewStoryStore(db *DB) *StoryStore {
	return &StoryStore{db: db}
}

// scanStory reads a Story from a row scanner, parsing SQLite TEXT timestamps.
func scanStory(row interface{ Scan(...any) error }) (*domain.Story, error) {
	var (
		s                    domain.Story
		createdAt, updatedAt string
	)
	if err := row.Scan(
		&s.ID, &s.EpicID, &s.Name, &s.Description, &s.ApprovalStatus,
		&createdAt, &updatedAt,
	); err != nil {
		return nil, err
	}
	var err error
	if s.CreatedAt, err = time.Parse(sqliteTimeFmt, createdAt); err != nil {
		return nil, fmt.Errorf("parse created_at: %w", err)
	}
	if s.UpdatedAt, err = time.Parse(sqliteTimeFmt, updatedAt); err != nil {
		return nil, fmt.Errorf("parse updated_at: %w", err)
	}
	return &s, nil
}

// Create inserts a new story or returns the existing one if (epicID, name) already exists.
// The created flag is true when a new row was inserted.
func (s *StoryStore) Create(epicID int, name, description string) (*domain.Story, bool, error) {
	res, err := s.db.conn.Exec(
		"INSERT OR IGNORE INTO stories (epic_id, name, description) VALUES (?, ?, ?)",
		epicID, name, description,
	)
	if err != nil {
		return nil, false, fmt.Errorf("create story: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return nil, false, fmt.Errorf("rows affected: %w", err)
	}
	story, err := s.GetByName(epicID, name)
	if err != nil {
		return nil, false, fmt.Errorf("fetch after create: %w", err)
	}
	return story, affected > 0, nil
}

// GetByName retrieves a story by its unique (epicID, name) pair.
// It returns domain.ErrNotFound when no matching story exists.
func (s *StoryStore) GetByName(epicID int, name string) (*domain.Story, error) {
	row := s.db.conn.QueryRow(
		"SELECT id, epic_id, name, description, approval_status, created_at, updated_at FROM stories WHERE epic_id = ? AND name = ?",
		epicID, name,
	)
	story, err := scanStory(row)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, fmt.Errorf("story %q (epic %d): %w", name, epicID, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("get story: %w", err)
	}
	return story, nil
}

// List returns all stories for the given epicID ordered by creation time.
func (s *StoryStore) List(epicID int) ([]domain.Story, error) {
	rows, err := s.db.conn.Query(
		"SELECT id, epic_id, name, description, approval_status, created_at, updated_at FROM stories WHERE epic_id = ? ORDER BY created_at",
		epicID,
	)
	if err != nil {
		return nil, fmt.Errorf("list stories: %w", err)
	}
	defer rows.Close()

	var stories []domain.Story
	for rows.Next() {
		story, err := scanStory(rows)
		if err != nil {
			return nil, fmt.Errorf("scan story: %w", err)
		}
		stories = append(stories, *story)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate stories: %w", err)
	}
	return stories, nil
}

// Update changes the description of an existing story.
// It returns domain.ErrNotFound when no matching story exists.
func (s *StoryStore) Update(epicID int, name, description string) (*domain.Story, error) {
	res, err := s.db.conn.Exec(
		"UPDATE stories SET description = ?, updated_at = datetime('now') WHERE epic_id = ? AND name = ?",
		description, epicID, name,
	)
	if err != nil {
		return nil, fmt.Errorf("update story: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return nil, fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return nil, fmt.Errorf("story %q (epic %d): %w", name, epicID, domain.ErrNotFound)
	}
	return s.GetByName(epicID, name)
}

// Delete removes a story by (epicID, name).
// It returns domain.ErrNotFound when no matching story exists.
func (s *StoryStore) Delete(epicID int, name string) error {
	res, err := s.db.conn.Exec(
		"DELETE FROM stories WHERE epic_id = ? AND name = ?",
		epicID, name,
	)
	if err != nil {
		return fmt.Errorf("delete story: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return fmt.Errorf("story %q (epic %d): %w", name, epicID, domain.ErrNotFound)
	}
	return nil
}

// UpdateApprovalStatus sets the approval status for a story.
// It returns domain.ErrNotFound when no matching story exists.
func (s *StoryStore) UpdateApprovalStatus(epicID int, name string, status domain.ApprovalStatus) error {
	res, err := s.db.conn.Exec(
		"UPDATE stories SET approval_status = ?, updated_at = datetime('now') WHERE epic_id = ? AND name = ?",
		status, epicID, name,
	)
	if err != nil {
		return fmt.Errorf("update approval status: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return fmt.Errorf("story %q (epic %d): %w", name, epicID, domain.ErrNotFound)
	}
	return nil
}

File: apps/orchestrator/src/db/stores/story-store.ts
import type { Database } from "bun:sqlite";
import type { Story, StorySummary } from "../../domain/entities.ts";
import type { ApprovalStatus } from "../../domain/enums.ts";
import { ErrNotFound, ErrAlreadyExists } from "../../domain/errors.ts";

export class StoryStore {
  constructor(private db: Database) {}

  create(epicId: number, name: string, description: string): Story {
    const stmt = this.db.prepare(
      "INSERT OR IGNORE INTO stories (epic_id, name, description) VALUES (?, ?, ?)",
    );
    const result = stmt.run(epicId, name, description);

    if (result.changes === 0) {
      throw ErrAlreadyExists("story", name);
    }

    return this.getByName(epicId, name);
  }

  getByName(epicId: number, name: string): Story {
    const row = this.db
      .query(
        "SELECT id, epic_id, name, description, approval_status, created_at, updated_at FROM stories WHERE epic_id = ? AND name = ?",
      )
      .get(epicId, name) as Story | null;

    if (!row) {
      throw ErrNotFound("story", name);
    }

    return row;
  }

  list(epicId: number): Story[] {
    return this.db
      .query(
        "SELECT id, epic_id, name, description, approval_status, created_at, updated_at FROM stories WHERE epic_id = ? ORDER BY created_at",
      )
      .all(epicId) as Story[];
  }

  update(epicId: number, name: string, description: string): Story {
    const stmt = this.db.prepare(
      "UPDATE stories SET description = ?, updated_at = datetime('now') WHERE epic_id = ? AND name = ?",
    );
    const result = stmt.run(description, epicId, name);

    if (result.changes === 0) {
      throw ErrNotFound("story", name);
    }

    return this.getByName(epicId, name);
  }

  delete(epicId: number, name: string): void {
    const stmt = this.db.prepare(
      "DELETE FROM stories WHERE epic_id = ? AND name = ?",
    );
    const result = stmt.run(epicId, name);

    if (result.changes === 0) {
      throw ErrNotFound("story", name);
    }
  }

  updateApprovalStatus(epicId: number, name: string, status: ApprovalStatus | "none"): void {
    const stmt = this.db.prepare(
      "UPDATE stories SET approval_status = ?, updated_at = datetime('now') WHERE epic_id = ? AND name = ?",
    );
    const result = stmt.run(status, epicId, name);

    if (result.changes === 0) {
      throw ErrNotFound("story", name);
    }
  }

  listSummaries(epicId: number): StorySummary[] {
    const query = `
      SELECT
        s.id, s.epic_id, s.name, s.description, s.approval_status, s.created_at, s.updated_at,
        COALESCE(tc.total, 0) AS task_count,
        COALESCE(tc.completed, 0) AS completed_task_count,
        COALESCE(tc.failed, 0) AS failed_task_count
      FROM stories s
      LEFT JOIN (
        SELECT story_id,
          COUNT(*) AS total,
          COUNT(CASE WHEN status = 'completed' THEN 1 END) AS completed,
          COUNT(CASE WHEN status = 'failed' THEN 1 END) AS failed
        FROM tasks GROUP BY story_id
      ) tc ON tc.story_id = s.id
      WHERE s.epic_id = ?
      ORDER BY s.created_at
    `;

    return this.db.query(query).all(epicId) as StorySummary[];
  }
}
